### PR TITLE
Fix up building and pushing OCI Images

### PR DESCRIPTION
**Pull request by rhatdan**
## Summary by Sourcery

Refactor the OCI image building and pushing process to separate the build and manifest creation steps, and fix the push command to handle image names correctly.

Bug Fixes:
- Fix the push command to correctly handle source and target image names, ensuring proper image conversion and manifest creation.

Enhancements:
- Refactor the OCI image building process to separate the build and manifest creation steps, improving code clarity and maintainability.

### CODE REVIEW

File: ramalama/oci.py [hunk 1]
```diff
@@ -116,37 +116,46 @@
...
+        try:
+            run_cmd(
+                [self.conman, "manifest", "rm", target], ignore_stderr=True, stdout=None, debug=args.debug
+            )
+        except subprocess.CalledProcessError as e:
+            pass
+        imageid = self._build(source, target, args)
+        imageid = self._create_manifest(target, imageid, args)
+
     def push(self, source, args):
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply ramalama/oci.py:143]: This is a good point, but I have a follow on PR where it will use the _convert function, separately.

**Comment by rhatdan** [comment]
@nalind PTAL

**Review by saschagrunert — APPROVED** [review]

File: ramalama/oci.py [hunk 2]
```diff
@@ -116,37 +156,54 @@ def _build(self, source, target, args):
...
+            "--annotation",
+            f"{annotations.AnnotationModel}=true",
+            "--annotation",
+            f"{ocilabeltype}=''",
+            "--annotation",
+            f"{annotations.AnnotationTitle}=args.SOURCE",
+            target,
+        ]
+        run_cmd(cmd_args, stdout=None, debug=args.debug)
+        run_cmd([self.conman, "manifest", "add", target, imageid], stdout=None, debug=args.debug)
```
Comment (saschagrunert) [inline ramalama/oci.py:180]: Can/should we set the artifact-type here?

Comment (rhatdan) [reply ramalama/oci.py]: I think I will just list all manifest lists as AI Models via ramalama list when using Docker, and then I can handle annotations differently with Podman.

**Comment by rhatdan** [comment]
With the new functionality we have to force the use of podman 5.0 for creation of OCI Images.  Anything less is not working properly.  This means we need Ubuntu 24.10 or newer, which kind of stinks.

**Comment by rhatdan** [comment]
@ericcurtin @saschagrunert @lsm5 I am thinking the tests are never running because github can not find VMs with the newer version of Ubuntu on it.  The issue I have is that the features added in this PR require a newer version of Podman to run "Podman 5.*" and ubuntu 24.04 only has Podman 4.9*

I can make RamaLama work with the Docker path for Podman < 5.0 but we will not end up with annotations in the OCI images.

**Comment by lsm5** [comment]
> @ericcurtin @saschagrunert @lsm5 I am thinking the tests are never running because github can not find VMs with the newer version of Ubuntu on it. The issue I have is that the features added in this PR require a newer version of Podman to run "Podman 5._" and ubuntu 24.04 only has Podman 4.9_

Yup, that's where testing-farm / TMT work much better. You get the latest fedora and centos OOTB and can be triggered via packit. I see there's a testing-farm job for centos-stream already.

RE: pushing images, IIUC, tmt folks are working on secrets handling feature, so if pushing can be handled separately, TMT should be able to handle everything else.

**Comment by lsm5** [comment]
> Yup, that's where testing-farm / TMT work much better. You get the latest fedora and centos OOTB and can be triggered via packit. I see there's a testing-farm job for centos-stream already.

NVM, I see there are other testing-farm jobs already. Let me check with testing-farm people about latest on secrets support.

**Comment by lsm5** [comment]
> NVM, I see there are other testing-farm jobs already. Let me check with testing-farm people about latest on secrets support.

secrets support seems to be implemented in testing-farm but not yet deployed. https://issues.redhat.com/browse/TFT-2472 . So, I guess hopefully soon.

File: docs/ramalama.1.md [hunk 3]
```diff
@@ -19,7 +19,7 @@ Running in containers eliminates the need for users to configure the host system
 
 RamaLama pulls AI Models from model registries. Starting a chatbot or a rest API service from a simple single command. Models are treated similarly to how Podman and Docker treat container images.
 
-When both Podman and Docker are installed, RamaLama defaults to Podman, The `RAMALAMA_CONTAINER_ENGINE=docker` environment variable can override this behavior. When neither are installed RamaLama attempts to run the model with software on the local system.
+When both Podman and Docker are installed, RamaLama defaults to Podman, The `RAMALAMA_CONTAINER_ENGINE=docker` environment variable can override this behaviour. When neither are installed RamaLama attempts to run the model with software on the local system.
```
Comment (nalind) [inline docs/ramalama.1.md:22]: Both are correct depending on where you are.

File: ramalama/oci.py [hunk 4]
```diff
@@ -116,37 +183,80 @@ def _build(self, source, target, args):
...
-            [self.conman, "build", "-t", target, "-f", containerfile.name, contextdir], stdout=None, debug=args.debug
+        imageid = (
+            run_cmd([self.conman, "build", "--no-cache", "-q", "-f", containerfile.name, contextdir], debug=args.debug)
+            .stdout.decode("utf-8")
+            .strip()
         )
+        return imageid
+
+    def tag(self, imageid, target, args):
+        # Create manifest list for target with imageid
```
Comment (nalind) [inline ramalama/oci.py:194]: Copy/pasted comment?

File: docs/ramalama.1.md (hunk 3 above)
Comment (ericcurtin) [reply docs/ramalama.1.md:22]: Yeah I've noticed this a few times, UK Vs US English. IMO just merge, because both are correct !

**Pull request merged**